server/src/main/java/org/elasticsearch/Version.java

```diff
@@ -285,6 +286,14 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         static final List<Version> DECLARED_VERSIONS = Collections.unmodifiableList(getDeclaredVersions(Version.class));
     }
 
+    // lazy initialized because we don't yet have the declared versions ready when instantiating the cached Version
+    // instances
+    private Version minCompatVersion;
```

Comment:
technically shouldn't we declare these as `volatile`?

Reply:
> technically shouldn't we declare these as volatile?

I figured no because we don't actively deduplicate this value so if we recompute it a few times (and realistically this is probably almost never going to happen) at startup it doesn't hurt but we pay the overhead in method size for `volatile` forever. But this was coded wrong even under that assumption now that I look at it again. I pushed https://github.com/elastic/elasticsearch/pull/62216/commits/153763ada87eb9392ad7198d04816ecf641658da so that it's safe even if out of order reads were to ever occur.